src/we_mac.c

```diff
@@ -1331,12 +1331,87 @@ static int we_hmac_set_priv_key(EVP_PKEY *pkey, const unsigned char *priv,
         ASN1_OCTET_STRING_free(asn1);
     }
 
-    WOLFENGINE_LEAVE(WE_LOG_MAC, "we_hmac_set_priv_key", ret);
+    WOLFENGINE_LEAVE(WE_LOG_MAC, "we_hmac_pkey_asn1_set_priv_key", ret);
+
+    return ret;
+}
+
+/**
+ * Get the private key as a byte buffer from the passed in EVP_PKEY.
+ *
+ * @param  pkey  [in]   EVP_PKEY to get private key buffer from.
+ * @param  priv  [out]  Output byte buffer to hold private key.
+ * @param  len   [out]  Holds the length of priv.
+ * @returns  1 on success, 0 on failure.
+ */
+static int we_hmac_pkey_asn1_get_priv_key(const EVP_PKEY *pkey,
+     unsigned char *priv, size_t *len)
+{
+    int ret = 1;
+    const unsigned char* privTmp;
+
+    WOLFENGINE_ENTER(WE_LOG_MAC, "we_hmac_pkey_asn1_get_priv_key");
+
+    if (pkey == NULL || len == NULL) {
+        WOLFENGINE_ERROR_MSG(WE_LOG_MAC, "Bad argument.");
+        ret = 0;
+    }
+
+    if (ret == 1) {
+        privTmp = EVP_PKEY_get0_hmac(pkey, len);
+        if (privTmp == NULL) {
+            ret = 0;
+        }
+        else {
+            if (priv != NULL) {
+                XMEMCPY(priv, privTmp, *len);
+            }
+        }
+    }
+
+    WOLFENGINE_LEAVE(WE_LOG_MAC, "we_hmac_pkey_asn1_get_priv_key", ret);
 
     return ret;
 }
 #endif
 
+/**
+ * Compare the public keys (ASN1_OCTET_STRINGs) held in a and b.
+ *
+ * @param  a  [in]  First key for comparison.
+ * @param  b  [in]  Second key for comparison.
+ * @returns  1 if the keys match and 0 if they don't.
+ */
+static int we_hmac_pkey_pub_cmp(const EVP_PKEY* a, const EVP_PKEY* b)
+{
+    int ret = 1;
+    ASN1_OCTET_STRING* aString;
+    ASN1_OCTET_STRING* bString;
+
+    WOLFENGINE_ENTER(WE_LOG_MAC, "we_hmac_pkey_pub_cmp");
+
+    if (a == NULL || b == NULL) {
+        WOLFENGINE_ERROR_MSG(WE_LOG_MAC, "Bad argument.");
+        ret = 0;
```

Comment:
-2 on error?

Reply:
Is this a suggestion in order to distinguish mismatches from errors, or is -2 on error documented somewhere? I'm good with the change either way, just curious.

Reply:
I checked OpenSSL and they are returning -2 for this type of error in related functions: eckey_pub_cmp().

Reply:
The OpenSSL code is in hm_ameth.c:

```
static int hmac_pkey_public_cmp(const EVP_PKEY *a, const EVP_PKEY *b)
{
    return ASN1_OCTET_STRING_cmp(EVP_PKEY_get0(a), EVP_PKEY_get0(b));
}
```

It doesn't bother checking the arguments.

Reply:
In theory, the parameters will always be good as the caller will have validated them.
But, for future proofing, we should check.

EVP_PKEY_cmp returns the result from the call to a->ameth->pub_cmp().
Previously in the function it returns -1 when types don't match and -2 when no ameth on a.

Reply:
Modified it to return -2 on error.